File: java/graph/PossibleBipartition.java
package graph;

public class PossibleBipartition {
    /* 和判断二分图几乎一模一样 区别只在于一些细节
     * 比如给的data不是adjacent list or adjacent matrix
     * 只是一个dislike list 我们需要转化成标准的adjacent list 注意这是undirected 两边都要更新
     * 还有就是index从1开始 为了方便起见 我们也从1开始 memory的构建都多加一个
     */
    boolean[] visited;
    boolean[] colors;
    boolean answer = true;
    public boolean possibleBipartition(int n, int[][] dislikes) {
        visited = new boolean[n + 1];
        colors = new boolean[n + 1];
        List<Integer>[] graph = buildGraph(dislikes, n);
        for(int i = 1;i < n + 1;i++) {
            if(!visited[i]) traverse(graph, i);
        }
        return answer;
    }
    private void traverse(List<Integer>[] graph, int current) {
        if(!answer) return;
        visited[current] = true;
        for(int i: graph[current]) {
            if(!visited[i]){
                colors[i] = !colors[current];
                traverse(graph, i);
            }else{
                if(colors[i] == colors[current]){
                    // not bipartite
                    answer = false;
                    return;
                }
            }
            
        }
    }
    private List<Integer>[] buildGraph(int[][] dislikes, int n) {
        List<Integer>[] graph = new LinkedList[n + 1];
        for(int i = 1;i < n + 1;i++) {
            graph[i] = new LinkedList<>();
        }
        for(int[] pair: dislikes) {
            graph[pair[0]].add(pair[1]);
            graph[pair[1]].add(pair[0]);
        }
        return graph;
    }
}

File: c++/graph/possibleBipartition.cpp
/**
 * 886. Possible Bipartition
 * C++版本的实现 比较不同的是这里用了C++书中的BFS版本
 * 而且在实现过程中我自己合并了visited和colors 0,1,2代表visited和两种color
 * 这会需要我们在一开始给第一个element一个颜色 然后之后所有node的颜色都是由previous node定好的
*/
class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        vector<vector<int>> graph(n, vector<int>());
        buildGraph(dislikes, graph);
        vector<int> visited(n, 0);
        queue<int> q;
        for(int i = 0;i < n; ++i) {
            if(visited[i]) continue;
            q.push(i);
            visited[i] = 1;
            while(!q.empty()) {
                int current = q.front();
                q.pop();
                for(const int & next : graph[current]) {
                    if(!visited[next]) {
                        visited[next] = visited[current] == 1 ? 2 : 1;
                        q.push(next);
                    }else if(visited[next] == visited[current]) return false;
                }
            }
        }
        return true;
    }

    void buildGraph(vector<vector<int>>& edge_list, vector<vector<int>>& graph) {
        for(const vector<int>& edge: edge_list) {
            graph[edge[1]-1].push_back(edge[0]-1);
            graph[edge[0]-1].push_back(edge[1]-1);
        }
    }
};